source/common/network/io_socket_handle_impl.cc

```diff
@@ -517,11 +528,27 @@ Address::InstanceConstSharedPtr IoSocketHandleImpl::peerAddress() {
   return Address::addressFromSockAddr(ss, ss_len);
 }
 
-Event::FileEventPtr IoSocketHandleImpl::createFileEvent(Event::Dispatcher& dispatcher,
-                                                        Event::FileReadyCb cb,
-                                                        Event::FileTriggerType trigger,
-                                                        uint32_t events) {
-  return dispatcher.createFileEvent(fd_, cb, trigger, events);
+void IoSocketHandleImpl::initializeFileEvent(Event::Dispatcher& dispatcher, Event::FileReadyCb cb,
+                                             Event::FileTriggerType trigger, uint32_t events) {
+  ASSERT(file_event_ == nullptr, "Attempting to initialize two `file_event_` for the same "
+                                 "file descriptor. This is not allowed.");
+  file_event_ = dispatcher.createFileEvent(fd_, cb, trigger, events);
+}
+
+void IoSocketHandleImpl::activateFileEvents(uint32_t events) {
+  if (file_event_) {
+    file_event_->activate(events);
+  } else {
+    ENVOY_BUG(false, "Null file_event_");
```

Comment:
I probably missed the convo here, but it's hard for me to understand how the server work in any sane way if file_event_ is nullptr, same below. Not sure if you had any discussion here about this.

Reply:
Yeah I agree we are basically hosed if this happens, so seems like this should be a lot noisier if this actually happens. Do we have tests for these cases?

Reply:
It come from the discussion above and it is more in the spirit of defensive programming to avoid crashing here.

Reply:
It would be great to add a basic test that verifies that these two calls trigger ENVOY_BUG when file_event_ is null

Reply:
This should be straight forward to unit test in `io_socket_handle_impl_test`. Do you mind if I do this in a separate PR?

Reply:
Sure we can do this in a different PR, but in general if we have defensive logic like this we should have tests for it. In this particular case I'm not convinced things would work correctly at all (timeout, etc.) but I'm fine with it.